src/master-playlist-controller.js

```diff
@@ -429,15 +429,16 @@ export class MasterPlaylistController extends videojs.EventTarget {
       // This is because in LIVE, the player plays 3 segments from the end of the
       // playlist, and if `BUFFER_LOW_WATER_LINE` is greater than the duration availble
       // in those segments, a viewer will never experience a rendition upswitch.
-      if (!currentPlaylist.endList ||
-          // For the same reason as LIVE, we ignore the low water line when the VOD
-          // duration is below the max potential low water line
-          this.duration() < Config.MAX_BUFFER_LOW_WATER_LINE ||
-          // we want to switch down to lower resolutions quickly to continue playback, but
-          nextPlaylist.attributes.BANDWIDTH < currentPlaylist.attributes.BANDWIDTH ||
-          // ensure we have some buffer before we switch up to prevent us running out of
-          // buffer while loading a higher rendition.
-          forwardBuffer >= bufferLowWaterLine) {
+      if (nextPlaylist &&
+          (!currentPlaylist.endList ||
```

Comment:
It might even be better to switch this giant if statement up into `return` if statements to make it more readable/understandable.

```js
// If the playlist is live, then we want to not take low water line into account.
// This is because in LIVE, the player plays 3 segments from the end of the	      
// playlist, and if `BUFFER_LOW_WATER_LINE` is greater than the duration available	 
// in those segments, a viewer will never experience a rendition up-switch.
if (currentPlaylist.endList) {
  return;
}

// For the same reason as LIVE, we ignore the low water line when the VOD
// duration is below the max potential low water line
if (this.duration() > Config.MAX_BUFFER_LOW_WATER_LINE) {
  return;
}

// we want to switch down to lower resolutions quickly to continue playback, but
if (nextPlaylist.attributes.BANDWIDTH > currentPlaylist.attributes.BANDWIDTH) {
  return;
}

// ensure we have some buffer before we switch up to prevent us running out of
if (forwardBuffer < bufferLowWaterLine) {
  return
}

Reply:
I like this, but I think @gkatsev makes a good point about re-triggering the `bandwidthupdate` event in the same order as before, just so as not to inadvertently break anyone.

I think a good way of doing this would be to create a new function that just takes in the parameters to check for if we `shouldSwitchToMedia`, then we can have the best of both worlds and a much easier function to test.